Add default book title and authors

SavageAvocado opened the pull request:
**Information**

This PR fixes https://github.com/EssentialsX/Essentials/issues/4912.

**Details**

**Proposed fix:**
Fixes "invalid book tag" issue by setting a default author & title for written books.

**Environments tested:**

OS: Windows 10

Java version: 17.0.1

- [X] Most recent Paper version (1.18.2)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**

Before (Tested with EssentialsX 2.20.0)
https://youtu.be/D7pgr_nSkAI

After (Tested by using the compiled version with this commit)
https://youtu.be/MS0klLUz2A0

SavageAvocado committed:
Fixes "invalid book tag" issue by setting a default author & title for written books.

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java:
```diff
@@ -10,10 +10,7 @@
-import org.bukkit.DyeColor;
-import org.bukkit.FireworkEffect;
-import org.bukkit.Material;
+import org.bukkit.*;
```
Don't use wildcard imports. You can configure your IDE not to do this.

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java:
```diff
@@ -45,6 +42,8 @@
 import static com.earth2me.essentials.I18n.tl;
 
 public class MetaItemStack {
+    private static final String DEFAULT_BOOK_AUTHOR = Bukkit.getConsoleSender().getName();
```
Not sure I like this... definitely more of an opinionated thing but I'd much rather it default to the player requesting/spawning the book, rather than console. Either that or make this configurable.

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java:
```diff
@@ -216,11 +215,20 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
+            // This fix only applies to written books.
+            if (stack.getType() == WRITTEN_BOOK) {
+                // Written books require an author and a title. https://bugs.mojang.com/browse/MC-59153
+                meta.setAuthor(DEFAULT_BOOK_AUTHOR);
```
We should probably try to use author meta here if applicable, and only default when it is not available. See the Essentials Item Meta format for books [here](https://wiki.mc-ess.net/wiki/Item_Meta#Books) on the old wiki.

pop4959 reviewed (dismissed):
Some minor things to touch up on, but this is looking great!

SavageAvocado committed:
Removed wildcard imports, checks that the book doesn't already have an author and or title before setting defaults, and the default author is now the name of the player who spawned the book.

SavageAvocado committed:
Merge branch '2.x' into 4912-fix-invalid-book-tag

SavageAvocado committed:
Merge branch 'EssentialsX:2.x' into 4912-fix-invalid-book-tag

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java:
```diff
@@ -216,11 +217,22 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
+            if (stack.getType() == WRITTEN_BOOK) {
+                // Written books require an author and a title. https://bugs.mojang.com/browse/MC-59153
+                if (!meta.hasAuthor()) {
+                    meta.setAuthor(sender.getDisplayName());
```
Particular reason for this being display name rather than username?

mdcfe replied:
I would also prefer we used the username here.

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java:
```diff
@@ -216,11 +217,22 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
+                    meta.setAuthor(sender.getDisplayName());
+                }
+                if (!meta.hasTitle()) {
+                    final String title = StringUtils.capitalize(split[1].toLowerCase(Locale.ROOT));
```
Should this not format the same way as for title meta below?

mdcfe replied:
I think the formatting is fine - `book:` values are often written in lowercase for convenience.

This will need updating to use our `StringUtil` as we no longer depend on Aapche Commons.

SavageAvocado committed:
Merge branch 'EssentialsX:2.x' into 4912-fix-invalid-book-tag

SavageAvocado committed:
Merge branch 'EssentialsX:2.x' into 4912-fix-invalid-book-tag

SavageAvocado committed:
Default book author is now the player's name, instead of the player's display name.
Written books from signs will also work without "*Invalid book tag*" errors now.

SavageAvocado committed:
Removed stray "org.bukkit.entity.Player" import.

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java:
```diff
@@ -216,11 +217,21 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
+            if (stack.getType() == WRITTEN_BOOK) {
+                if (!meta.hasAuthor()) {
+                    // The sender can be null when this method is called from {@link  com.earth2me.essentials.signs.EssentialsSign#getItemMeta(ItemStack, String, IEssentials)}
+                    meta.setAuthor(sender == null ? Bukkit.getConsoleSender().getName() : sender.getPlayer().getName());
```
I'm not sure if we want to use the console sender name, or if it would be better to have a translation key like `bookDefaultAuthor` that servers/translators can override? cc @pop4959 @JRoy

JRoy replied:
Agree with the translation

pop4959 replied:
I see no reason not to, I think in all cases where a non-player is spawning books it's likely they would want to set the `author` meta anyways, but this seems good to do anyway.

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java:
```diff
@@ -286,9 +297,9 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
-                        Integer.valueOf(input.substring(1, 3), 16),
-                        Integer.valueOf(input.substring(3, 5), 16),
-                        Integer.valueOf(input.substring(5, 7), 16)));
+                            Integer.valueOf(input.substring(1, 3), 16),
+                            Integer.valueOf(input.substring(3, 5), 16),
+                            Integer.valueOf(input.substring(5, 7), 16)));
```
Undo this unnecessary diff

JRoy replied:
```suggestion
                        Integer.valueOf(input.substring(1, 3), 16),
                        Integer.valueOf(input.substring(3, 5), 16),
                        Integer.valueOf(input.substring(5, 7), 16)));
```

JRoy committed:
Discard changes to .idea/checkstyle-idea.xml

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/MetaItemStack.java

JRoy committed:
Merge branch '2.x' into 4912-fix-invalid-book-tag

JRoy committed:
changes

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/signs/EssentialsSign.java:
```diff
@@ -46,7 +46,7 @@ protected static boolean checkIfBlockBreaksSigns(final Block block) {
             return true;
         }
-        final BlockFace[] directions = new BlockFace[] {BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST};
+        final BlockFace[] directions = new BlockFace[]{BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST};
```
```suggestion
        final BlockFace[] directions = new BlockFace[] {BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST};
```

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/signs/EssentialsSign.java
